tools/GenerateDocFxStructure/Program.cs

```diff
@@ -0,0 +1,313 @@
+                _returnvalue = 1;
+                return;
+            }
+
+            o.DestinationFolder = Path.GetFullPath(o.DestinationFolder);
+            o.MediaFolder = Path.GetFullPath(o.MediaFolder);
+            o.SourceFolder = Path.GetFullPath(o.SourceFolder);
+            _options = o;
+
+            _message.Verbose($"Source folder     : {_options.SourceFolder}");
+            _message.Verbose($"Destination folder: {_options.DestinationFolder}");
+            _message.Verbose($"Media folder      : {_options.MediaFolder}");
+            _message.Verbose($"Verbose           : {_options.Verbose}");
+
+            if (!Directory.Exists(_options.SourceFolder))
+            {
+                _message.Error($"ERROR: Documentation folder '{_options.SourceFolder}' doesn't exist.");
+                _returnvalue = 1;
+                return;
+            }
+
+            // We will start to get all the files
+            foreach (string file in Directory.EnumerateFiles(_options.SourceFolder, "*.*", SearchOption.AllDirectories))
+            {
+                allFiles.Add(file.ToLowerInvariant());
+            }
+
+
+            // we start at the root to generate the TOC items
+            DirectoryInfo rootDir = new DirectoryInfo(_options.SourceFolder);
+            WalkDirectoryTree(rootDir);
+        }
+
+        /// <summary>
+        /// Main function going through all the folders, files and subfolders.
+        /// </summary>
+        /// <param name="folder">The folder to search.</param>
+        private static void WalkDirectoryTree(DirectoryInfo folder)
+        {
+            _message!.Verbose($"Processing folder {folder.FullName}");
+
+            // process MD files in this folder
+            ProcessFiles(folder);
+
+            // process other sub folders
+            DirectoryInfo[] subDirs = folder.GetDirectories();
+            foreach (DirectoryInfo dirInfo in subDirs)
+            {
+                WalkDirectoryTree(dirInfo);
+            }
+        }
+
+        /// <summary>
+        /// Get the list of the files in the current directory.
+        /// </summary>
+        /// <param name="folder">The folder to search.</param>
+        private static void ProcessFiles(DirectoryInfo folder)
+        {
+            _message!.Verbose($"Process {folder.FullName} for files.");
+
+            List<FileInfo> files = folder.GetFiles("*.md").OrderBy(f => f.Name).ToList();
+            if (files == null)
+            {
+                _message.Verbose($"No MD files found in {folder.FullName}.");
+                return;
+            }
+
+            foreach (FileInfo fi in files)
+            {
+                _message.Verbose($"Processing {fi.FullName}");
+                string content = File.ReadAllText(fi.FullName);
+
+                // first see if there are links in this file
+                Regex rxContent = new Regex(@"(\[{1}.*\]{1}\({1}.*\){1}?)");
```

Comment:
1. you can cache this and precompile for some speed (not sure how fast this currently is and if this will actually make any observable difference)
2. Might be worth to comment some example stuff it matches so you don't have to decode it when reading code 😄

Reply:
what's the question mark for?

```
(\[{1}.*\]{1}\({1}.*\){1}?)
                         ^
```

Reply:
should all `.*` be `.+`? or even `[^()\[\]]` or something along the lines?

Reply:
We've been trying multiple combinations on another project See [this](https://github.com/Ellerbach/docfx-companion-tools/tree/main/src/DocLinkChecker) and there's been couple of attempts to find the right way of writing the regex. It seems that after some time, the one in this code has been proven to work in all situations (with absolute files, relative, / or \ or url). 
To answer you question, the last ? is to match the previous element only once.
And, yes the regex can be precompiled or even just as a private readonly variable, that would optimize a bit the time of execuation. On my machine, without optimization, in verbose, it's about 2 seconds for the all repo. I can anyway optimize this a bit.
Current limitation is on image link like: `[![image](image)](anotherlink)` which are a bit more complicated to trap. Now, we don't have, if we know we have, it should be then checked within the extracted data to see if any other link (image) is in the extracted material.